Check only column name and type for async distributed blocks

azat opened the pull request:
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Correctly check structure of async distributed blocks.

Detailed description:
INSERT into Distributed with insert_distributed_sync=1 stores the
distributed batches on the disk for sending in background.

But types may be a little bit different for the Distributed and it's
underlying table, so the initiator need to know whether conversion is
required or not.

Before this patch those on disk distributed batches contains header,
which includes dumpStructure() for the block in that batch, however it
checks not only names and types and plus dumpStructure() is a debug
method.

So instead of storing string representation for the block header we
should store empty block in the file header (note, that we cannot store
the empty block not in header, since this will require reading all
blocks from file, due to some trickery of the readers interface).

Note, that this patch also contains tiny refactoring:
- s/header/distributed_header/

v1: dumpNamesAndTypes()
v2: dump empty block into the batch itself
v3: move empty block into the header
v4: improve coverage for the case when conversion should happen

Fixes: #21906
Fixes: #19947

vdimir commented on src/Storages/Distributed/DirectoryMonitor.cpp:
```diff
@@ -202,7 +208,7 @@ namespace
     void writeRemoteConvert(const DistributedHeader & header, RemoteBlockOutputStream & remote, ReadBufferFromFile & in, Poco::Logger * log)
     {
-        if (remote.getHeader() && header.header != remote.getHeader().dumpStructure())
+        if (remote.getHeader() && header.names_and_types != remote.getHeader().dumpNamesAndTypes())
```
Does sending and comparing string representations of block's names and types really gives gain instead of reading one block?
Actually `dumpNamesAndTypes` also looks like debug method like `dumpStructure`. 

Maybe we can send header on native format?

filimonov commented:
I think it should be backported to 21.3. So maybe mark it a bugfix?

azat committed:
Compare Block itself for distributed async INSERT batches

INSERT into Distributed with insert_distributed_sync=1 stores the
distributed batches on the disk for sending in background.

But types may be a little bit different for the Distributed and it's
underlying table, so the initiator need to know whether conversion is
required or not.

Before this patch those on disk distributed batches contains header,
which includes dumpStructure() for the block in that batch, however it
checks not only names and types and plus dumpStructure() is a debug
method.

So instead of storing string representation for the block header we
should store empty block in the file header (note, that we cannot store
the empty block not in header, since this will require reading all
blocks from file, due to some trickery of the readers interface).

Note, that this patch also contains tiny refactoring:
- s/header/distributed_header/

v1: dumpNamesAndTypes()
v2: dump empty block into the batch itself
v3: move empty block into the header

azat commented:
>I think it should be backported to 21.3. So maybe mark it a bugfix?

Well the only thing that will happen is one extra conversion (for LowCardinality columns for example), so I'm not sure that it worth it (although if 21.3 is LTS, then it make sense I guess)

azat commented:
>[AST fuzzer (ASan) — Fuzzer failed (107). See the logs.](https://clickhouse-test-reports.s3.yandex.net/22325/8e7fc45f385d622cb30d11c7476d3e2c4d08971a/fuzzer_asan/report.html#fail1)

`.sql` test was renamed to `.sh` and it breaks fuzzer, should be fixed by #22542

azat commented:
>[Yandex synchronization check (only for Yandex employees)](https://a.yandex-team.ru/review/1712540)
>[Yandex third-party checks (only for Yandex employees)](https://a.yandex-team.ru/review/1712540)

Any clue?

vdimir commented on src/Storages/Distributed/DistributedBlockOutputStream.cpp:
```diff
@@ -679,7 +679,13 @@ void DistributedBlockOutputStream::writeToShard(const Block & block, const std::
             writeVarUInt(block.rows(), header_buf);
             writeVarUInt(block.bytes(), header_buf);
-            writeStringBinary(block.cloneEmpty().dumpStructure(), header_buf);
+            writeStringBinary(block.cloneEmpty().dumpStructure(), header_buf); /// obsolute
```
```suggestion
            writeStringBinary(block.cloneEmpty().dumpStructure(), header_buf); /// obsolete
```

vdimir commented:
> > [Yandex synchronization check (only for Yandex employees)](https://a.yandex-team.ru/review/1712540)
> > [Yandex third-party checks (only for Yandex employees)](https://a.yandex-team.ru/review/1712540)
> 
> Any clue?

Same problem with master

azat commented:
>[ClickHouse build check — 0/11 builds are OK](https://clickhouse-builds.s3.yandex.net/22325/ba238709d6a1992ea21d8ad25804c6222a5d0e49/clickhouse_build_check/report.html)

Some CI issues, rebased against upstream (w/ your typo fix), previous HEAD is https://github.com/ClickHouse/ClickHouse/pull/22325/commits/e86469647a48d321161be389f3dcd145326ea31a

azat commented:
https://clickhouse-test-reports.s3.yandex.net/22325/7ccd784acb5554b6f39f4fdc72028becf5dfee55/stress_test_(memory).html#fail1 - #22592

azat commented:
>[ClickHouse build check — 14/15 builds are OK](https://clickhouse-builds.s3.yandex.net/22325/27d4fbd13b7c6eeade74a84ab0813c2fd1962eaf/clickhouse_build_check/report.html)

`Build failed without build log. It's inner CI problem. Maybe something wrong with docker or host. Just wait for build restart.` looks unrelated, but I did not see that build fails in upstream/master, is it really CI problem?

vdimir commented:
> is it really CI problem?

yes

millin commented:
Why wasn't it backported to 21.3 LTS?

azat commented:
>Why wasn't it backported to 21.3 LTS?

In my opinion it is not that significant for backporting (since it just do one extra conversion in some cases), however since 21.3 is LTS maybe it worth doing, @vdimir thoughts?

vdimir commented:
> Why wasn't it backported to 21.3 LTS?

I also think that if it is not bug fix, doesn't affect to user, but removes overhead in some cases, so backporting not necessary.

filimonov commented:
It spams in err.log which is monitored by many setups.

Especially important when you push logs to some external log monitoring system, where you pay per logs amount.

vdimir commented:
> It spams in err.log which is monitored by many setups.
> 
> Especially important when you push logs to some external log monitoring system, where you pay per logs amount.

Ok, I'll backport it
